### DISCUSSION TITLE
How to output chemical reaction rates?

### DISCUSSION THREAD

**Discussion by CAIROCharlie** [AMReX-Combustion/PelePhysics#511] [Q&A]
Hello,

I want to output the time evolution of chemical reaction rates for each reaction in my PeleLMeX case. How can I accomplish this?

Thank you very much.

**Comment by baperry2** [comment]
It is not possible to output individual reaction rates per reaction in PeleLMeX and I do not believe that you could work around this. The individual reaction rates are computed internally within PelePhysics code and then the net production rates of each species are computed and used within the chemistry substepping routines that integrate the chemical system over the CFD timestep. The individual reaction rates are never directly exposed in PeleLMeX.

If you want the reaction rates of individual reactions, your best bet may be save/plot mass fractions, temperature, and pressure and  recompute them in postprocessing using Cantera (there are Cantera format mechanisms available for all the mechanisms in PelePhysics). The _instantaneous_ reaction rates computed by Cantera should exactly match those used internally within PeleLMeX for a given thermochemical state and might be suitable depending on your objectives. It should be noted that while this would be an exact reproduction of the reaction rates at that instant, you could not use this approach to integrate the reaction system forward and match the PeleLMeX results, as the internal integration of the chemical system includes a forcing term from advection/diffusion that you would not have access to.

There are several ways to output the species net production rates over the CFD timestep if that will meet your needs, including outputting them in the plot files (`peleLM.plot_react = 1`), as well as tracking the integrated net change in species using the species balance temporal or using the Diagnostic capability to output slices or PDFs as described in the [documentation](https://amrex-combustion.github.io/PeleLMeX/manual/html/LMeXControls.html#run-time-diagnostics).

**Reply by CAIROCharlie** [reply]
Thank you very much for your help! I will give it a try.
